### PR TITLE
Kernel+LibThreading: Allow dead threads to be joined, and generally overhaul thread behavior

### PR DESCRIPTION
**Pull request by kleinesfilmroellchen**
## Part 1: yes, you can join a dead thread, you should be able to, POSIX tells you to, I checked
Joining dead threads should be allowed for two main reasons:
- Thread join behavior should not be racy when a thread is joined and exiting at roughly the same time. This is common behavior when threads are given a signal to end (meaning they are going to exit ASAP) and then joined.
- POSIX requires that exited threads are joinable. This means that we're more spec compliant here without hacky userland workarounds.

The behavior is still well-defined; e.g. it doesn't allow a dead detached thread to be joined or a thread to be joined more than once.

## Part 2: Make `Threading::Thread` thread-safe (lol) and well-behaved

TL;DR: also allow joining dead threads, just use one well-defined state variable and also crash whenever the user mishandles threads (as that usually leads to issues down the line + concurrency is hard)

## Part 3: Thread tests that are enabled and actually pass

(yay)

### CODE REVIEW

File: Kernel/Thread.h [hunk 1]
```diff
@@ -824,7 +824,12 @@ class Thread
             return EDEADLK;
 
         SpinlockLocker lock(m_lock);
-        if (!m_is_joinable || state() == Thread::State::Dead)
+
+        // Joining dead threads is allowed for two main reasons:
+        // - Thread join behavior should not be racy when a thread is joined and exiting at roughly the same time.
+        //   This is common behavior when threads are given a signal to end (meaning they are going to exit ASAP) and then joined.
+        // - POSIX requires that exited threads are joinable (at least, there is no language in the specification forbidding it).
+        if (!m_is_joinable || state() == Thread::State::Invalid)
```
Comment (ADKaster) [inline Kernel/Thread.h:832]: This is a bit different from the recommendations in the Rationale section of Dr. Posix:

> If an implementation detects that the value specified by the thread argument to pthread_join() does not refer to a joinable thread, it is recommended that the function should fail and report an [EINVAL] error.
>
> If an implementation detects that the value specified by the thread argument to pthread_join() refers to the calling thread, it is recommended that the function should fail and report an [EDEADLK] error.
> 
> If an implementation detects use of a thread ID after the end of its lifetime, it is recommended that the function should fail and report an [ESRCH] error.

https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_join.html

It sounds like it's recommending that we check the thread state here: if it's dead, we should return ESRCH. if it's not a joinable thread, we should return EINVAL.

Comment (ADKaster) [reply Kernel/Thread.h:832]: If a thread is dead or dying, we shouldn't block the calling thread until it gets reaped by the Finalizer thread. Not sure if this implementation will do that, but it seems like the posix recommendation makes sense to me.

Comment (kleinesfilmroellchen) [reply Kernel/Thread.h:832]: Hmmm interesting, I either had an old version of the POSIX standard open or missed that last paragraph; I was confident I scanned this page several times for anything relating to this. I'm still asking myself how that makes any sense; if I'm joining a finite-lifespan thread, it always (given enough "allowed" scheduler unfairness) has the possibility of exiting before I can call join(). We should get ESRCH left and right with normal "do one task then exit" threads, right?

Comment (kleinesfilmroellchen) [reply Kernel/Thread.h:832]: Either way, I think will add another commit that ignores ESRCH in our Core::Thread which doesn't have to adhere to POSIX semantics.

Comment (kleinesfilmroellchen) [reply Kernel/Thread.h:832]: See Discord; after re-checking I think the new behavior is correct.

**Comment by kleinesfilmroellchen** [comment]
Drafting because this is a yakchain of things we need to fix in the userland thread implementation.

File: Tests/LibThreading/TestThread.cpp [hunk 2]
```diff
@@ -29,10 +28,32 @@ TEST_CASE(threads_can_detach)
...
+}
+
+TEST_CASE(join_dead_thread)
+{
+    auto thread = Threading::Thread::construct([&]() { return 0; });
+    thread->start();
+    // The thread should have exited by then.
+    usleep(40 * 1000);
+
+    auto join_result = thread->join<int*>();
```
Comment (ADKaster) [inline Tests/LibThreading/TestThread.cpp:55]: this being int* when you returned an int is pretty sus. Should join use reinterpret_cast instead of static_cast?

**Review by ADKaster — CHANGES_REQUESTED** [review]
Nice cleanup! I have a few comments though.

File: Userland/Libraries/LibThreading/Thread.cpp [hunk 3]
```diff
@@ -56,16 +58,51 @@ String Thread::thread_name() const { return m_thread_name; }
...
 void Thread::start()
 {
+    VERIFY(!is_started());
+
+    // Set this first so that the other thread starts out seeing m_state == Running.
+    m_state = Threading::ThreadState::Running;
+
     int rc = pthread_create(
         &m_tid,
         nullptr,
```
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Userland/Libraries/LibThreading/Thread.cpp:86]: Hmm, yeah we would need an internal flag for when the user calls detach before start. I'll add a FIXME.

File: Tests/LibThreading/TestThread.cpp [hunk 4]
```diff
@@ -10,7 +10,6 @@
 
 // FIXME: Enable these tests once they work reliably.
```
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Tests/LibThreading/TestThread.cpp:11]: 🤦

File: Tests/LibThreading/TestThread.cpp (hunk 2 above)
Comment (kleinesfilmroellchen) [reply Tests/LibThreading/TestThread.cpp:55]: I think this is just to discard the value, as the 0 is for nullptr afaik. I will have a look next week once I have time.

**Comment by AtkinsSJ** [comment]
This has conflicts.

File: Userland/Libraries/LibThreading/Thread.cpp [hunk 5]
```diff
@@ -25,10 +25,12 @@ Thread::Thread(Function<intptr_t()> action, StringView thread_name)
 
 Thread::~Thread()
 {
-    if (m_tid && !m_detached) {
-        dbgln("Destroying thread \"{}\"({}) while it is still running!", m_thread_name, m_tid);
+    if (needs_to_be_joined()) {
+        dbgln("Destroying {} while it is still running undetached!", *this);
         [[maybe_unused]] auto res = join();
     }
+    if (m_state == ThreadState::Detached)
```
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply Userland/Libraries/LibThreading/Thread.cpp:32]: Yeah it is, a detached thread will cause the function to lose its parent or whatever, so it falls over and dies

**Review by ADKaster — APPROVED** [review]
